vit_jax/train.py

```diff
@@ -13,195 +13,215 @@
+  variables = jax.jit(init_model, backend='cpu')()
+
+  pretrained_path = os.path.join(config.pretrained_dir,
+                                 f'{config.model.name}.npz')
+  if not tf.io.gfile.exists(pretrained_path):
+    raise ValueError(
+        f'Could not find "{pretrained_path}" - you can download models from '
+        '"gs://vit_models/imagenet21k" or directly set '
+        '--config.pretrained_dir="gs://vit_models/imagenet21k".')
   params = checkpoint.load_pretrained(
       pretrained_path=pretrained_path,
-      init_params=params,
-      model_config=models.CONFIGS[args.model],
-      logger=logger)
+      init_params=variables['params'],
+      model_config=config.model)
+
+  total_steps = config.total_steps
+  lr_fn = utils.create_learning_rate_schedule(total_steps, config.base_lr,
+                                              config.decay_type,
+                                              config.warmup_steps)
 
-  # pmap replicates the models over all TPUs/GPUs
-  vit_fn_repl = jax.pmap(VisionTransformer.call)
-  update_fn_repl = make_update_fn(VisionTransformer.call, args.accum_steps)
+  update_fn_repl = make_update_fn(
+      apply_fn=model.apply, accum_steps=config.accum_steps, lr_fn=lr_fn)
+  infer_fn_repl = jax.pmap(functools.partial(model.apply, train=False))
 
   # Create optimizer and replicate it over all TPUs/GPUs
   opt = momentum_clip.Optimizer(
-      dtype=args.optim_dtype, grad_norm_clip=args.grad_norm_clip).create(params)
-  opt_repl = flax_utils.replicate(opt)
+      dtype=config.optim_dtype,
+      grad_norm_clip=config.grad_norm_clip).create(params)
+
+  initial_step = 1
+  opt, initial_step = flax_checkpoints.restore_checkpoint(
+      workdir, (opt, initial_step))
+  logging.info('Will start/continue training at initial_step=%d', initial_step)
+
+  opt_repl = flax.jax_utils.replicate(opt)
 
   # Delete references to the objects that are not needed anymore
   del opt
   del params
 
-  def copyfiles(paths):
-    """Small helper to copy files to args.copy_to using tf.io.gfile."""
-    if not args.copy_to:
-      return
-    for path in paths:
-      to_path = os.path.join(args.copy_to, args.name, os.path.basename(path))
-      tf.io.gfile.makedirs(os.path.dirname(to_path))
-      tf.io.gfile.copy(path, to_path, overwrite=True)
-      logger.info(f'Copied {path} to {to_path}.')
-
-  total_steps = args.total_steps or (
-      input_pipeline.DATASET_PRESETS[args.dataset]['total_steps'])
-
   # Prepare the learning-rate and pre-fetch it to device to avoid delays.
-  lr_fn = hyper.create_learning_rate_schedule(total_steps, args.base_lr,
-                                              args.decay_type,
-                                              args.warmup_steps)
-  lr_iter = hyper.lr_prefetch_iter(lr_fn, 0, total_steps)
-  update_rngs = jax.random.split(
-      jax.random.PRNGKey(0), jax.local_device_count())
+  update_rng_repl = flax.jax_utils.replicate(jax.random.PRNGKey(0))
```

Comment:
This changes the semantics from different seed per device to the same seed on all devices, which I would expect may change results!

Reply:
Note code inside `update_fn()`:

```
    _, new_rng = jax.random.split(rng)
    # Bind the rng key to the device id (which is unique across hosts)
    # Note: This is only used for multi-host training (i.e. multiple computers
    # each with multiple accelerators).
    dropout_rng = jax.random.fold_in(rng, jax.lax.axis_index('batch'))
```

So while the training loop pattern is different, the random values for dropout should still be unique to every host as before, PTAL.

Reply:
looks good to me: we in fact switched to the exact same design internally.

Reply:
Sounds good, I did scroll down to check that, but seems like I completely missed it!